packages/saas-ui-auth/src/services/supabase.ts

```diff
@@ -6,103 +6,190 @@ import {
-    const [key, value] = part.split('=')
-    memo[key] = value
-    return memo
-  }, {})
-}
-
-export const createAuthService = (supabase: any): AuthProviderProps => {
+  const hash = window.location.hash.replace("#", "");
+  return hash.split("&").reduce<any>((memo, part) => {
+    const [key, value] = part.split("=");
+    memo[key] = value;
+    return memo;
+  }, {});
+};
+
+export const createAuthService = (
+  supabase: SupabaseClient<any, "public", any>
+): AuthProviderProps => {
   const onLogin = async (params: AuthParams, options?: AuthOptions) => {
-    const { user, error } = await supabase.auth.signIn(
-      params as unknown,
-      options
-    )
-
-    if (user) {
-      return user as User
-    } else if (error) {
-      throw error
+    function authenticate() {
+      const { email, password, provider, phone } = params;
+      if (email && password) {
+        return supabase.auth.signInWithPassword({
+          email,
+          password,
+        });
+      } else if (email) {
+        return supabase.auth.signInWithOtp({ email });
+      } else if (provider) {
+        return supabase.auth.signInWithOAuth({
+          provider: provider as Provider,
+        });
+      } else if (phone && password) {
+        return supabase.auth.signInWithPassword({ phone, password });
+      } else if (phone) {
+        return supabase.auth.signInWithOtp({ phone });
+      }
+      throw new Error("Could not find correct authentication method");
+    }
+    const resp = await authenticate();
+
+    if (resp.error) {
+      throw resp.error;
+    }
+    if (isOauthResponse(resp)) {
+      const userResp = await supabase.auth.getUser();
+      if (userResp.error) {
+        throw userResp.error;
+      }
+      return userResp.data.user;
     }
-  }
+    return resp.data.user;
+  };
 
   const onSignup = async (params: AuthParams, options?: AuthOptions) => {
-    const { email, password, ...data } = params
-    const { user, error } = await supabase.auth.signUp(
-      {
-        email,
-        password,
-      },
-      {
+    const { email, password, phone, captchaToken, emailRedirectTo, data } =
+      params;
+    if (!password) {
+      throw new Error("Need to provide password to signUp");
+    }
+
+    const resp = await supabase.auth.signUp({
+      email,
+      password,
+      phone,
+      options: {
+        captchaToken,
+        emailRedirectTo,
         data,
-        ...options,
-      }
-    )
+      },
+    });
 
-    if (user) {
-      return user as User
-    } else if (error) {
-      throw error
+    if (resp.error) {
+      throw resp.error;
     }
-  }
+    return resp.data.user;
+  };
 
   const onVerifyOtp = async (params: AuthParams) => {
-    const { session, error } = await supabase.auth.verifyOTP(params)
+    // redirectTo & captcha are part of the options, but not for AuthParams, maybe add there?
+    const { email, phone, otp, type, redirectTo, captchaToken } = params;
```

Comment:
use an extra options parameter, just like the other handlers.

this is the function signature:
```ts
  onVerifyOtp?: (
    params: AuthParams,
    options?: AuthOptions
  ) => Promise<boolean | undefined | null>
  ```

Reply:
Did this in all the other calls now as well. Extended the AuthOptions type to match supabase better.